### ISSUE TITLE
1 plugin per language

### ISSUE THREAD

**Issue by angelozerr** [eclipse/wildwebdeveloper#26]
IMHO, I think org.eclipse.bluesky should be splitted in several plugins: one plugin per language.

**Comment by mickaelistria** [comment]
Can you please elaborate why? What are the existing use-cases for this?

**Comment by angelozerr** [comment]
> Can you please elaborate why? What are the existing use-cases for this?

 * there are a several configuration for a given language: language server, textmate, etc which is very time the same. I'm working on VSCode language-configuration.json and it will add new file configuration. So when we will want to add another language (like xml for instance), it will be easy to copy/paste an existing plugin (ex: org.eclipse.bluesky.json) and do find/replace json by xml, and that's all. In the future we could provide a wizard which generates a plugin for lsp4e. For CSS server, it manages notifications to support colorisation, to do that we need to defines a CSSLanguageServerInterface. it's like JSONLanguageServerInterface

 * I have added JSON Schema support in https://github.com/mickaelistria/eclipse-bluesky/issues/25 and it requires several new classes like JSONLanguageClient and JSONLanguageServerInterface. Thoses classes pollute the other languages.

**Comment by mickaelistria** [comment]
> it will be easy to copy/paste an existing plugin (ex: org.eclipse.bluesky.json) and do find/replace json by xml, and that's all. In the future we could provide a wizard which generates a plugin for lsp4e.

If this is with the goal of contributing to bluesky, then I don't get why multiple bundles make copy-paste easier than copy-pasting from and into the same bundle.

> I have added JSON Schema support in #25 and it requires several new classes like JSONLanguageClient and JSONLanguageServerInterface. Thoses classes pollute the other languages. For CSS server, it manages notifications to support colorisation, to do that we need to defines a CSSLanguageServerInterface. it's like JSONLanguageServerInterface

Why do you need separate bundles to contribute that? The current bundle could totally welcome those classes.

**Comment by angelozerr** [comment]
> Why do you need separate bundles to contribute that? The current bundle could totally welcome those classes.

I prefer having a plugin per language which configure the language instead having one plugin which does all (it will be become difficult when we will had custom behaviour for a language).

VSCode does that https://github.com/Microsoft/vscode/tree/master/extensions. Each extension have a textmate gramme file a language-configuration.json file etc. 

I would like to follow the same idea, but if you don't see the necesserity, I will stop to insist.

**Comment by mickaelistria** [comment]
I don't think it's a bad idea, I just don't have the impression this is high priority at the moment as it wouldn't deliver much added-value compared to the current state. That's why I was asking for a user-story to support this change to decide of how high I should place it in my todo-list ;)
But if this is very important to you and you feel ready to make a PR for it, that would be welcome!

**Comment by angelozerr** [comment]
>  I just don't have the impression this is high priority at the moment as it wouldn't deliver much added-value compared to the current state. 

I would like to do several PR like JSON Schema contribution, language configuration, xml support, etc and I'm afraid that bluesky will start be complex because there are a lot of configuration file.

That's why I have suggested to split into several plugins. I could do that if you wish but my main problem is about org.eclipse.bluesky\language-servers, I have not very understood how dependencies from https://github.com/mickaelistria/eclipse-bluesky/blob/master/org.eclipse.bluesky/language-servers/package.json is working (it doesn't work for me because I have windows).

In other words, I can split plugins, but I'm afraid to break the https://github.com/mickaelistria/eclipse-bluesky/blob/master/org.eclipse.bluesky/language-servers. Are you OK I do a PR and you check that language server https://github.com/mickaelistria/eclipse-bluesky/blob/master/org.eclipse.bluesky/language-servers/ is working?

**Comment by angelozerr** [comment]
@mickaelistria see my PR https://github.com/mickaelistria/eclipse-bluesky/pull/31 I have not splitted plugins but just reorganize declaration and package name

**Comment by angelozerr** [comment]
@mniewrzal seems wishing split of plugins per language https://github.com/eclipselabs/language-servers-for-eclipse/issues/2 no?

**Comment by mickaelistria** [comment]
One issue with splitting bundles is that since most LS do share the same npm modules, we'd either have to duplicate some modules or have to set up a strategy to share modules across LS from different bundles.

**Comment by mickaelistria** [comment]
That said, at the moment, the XML LS has no relationship with other LS and could be extracted to its own bundle.

**Comment by mickaelistria** [comment]
With https://github.com/eclipse/wildwebdeveloper/commit/470b0738862c273b454cb231c50b6ea032b44c78 , I moved the XML support to its own bundle as it doesn't share node_modules with other LS.

**Comment by jabby** [comment]
Thanks.

**Comment by mickaelistria** [comment]
To avoid duplication of node.js modules, as explained here and there, we cannot easily do that without introducing much more maintenance complexity and much more disk usage (on client side and in bandwidth).
I'm closing for now. We'll reopen finer grain ticket if there is urge to "extract" some LS at some point.
